### PR TITLE
[Merged by Bors] - Add automatic docs deployment to GitHub Pages

### LINKED ISSUE #407: Publish API Documentation for Master Branch
I was just thinking that it would be useful to have the API documentation for the master branch published to GitHub pages somewhere.

I was going to explain to @lee-orr, a beginner Rustacean, what changes were going to need to happen for #397, but I noticed that some of the API that I was going to point to had not been released yet, which meant that I can't point to the documentation for it. I didn't end up needing it in that case, but it seems like it could be useful.

### PR DESCRIPTION
**Pull request by james7132**
# Objective

Partially address #407 by setting up automated deployments of `bevy`'s API reference to GitHub Pages.

## Solution

Set up a GitHub Actions workflow that builds the docs on every push to `main` and pushes a new commit to a `gh-pages` (or `docs` branch).

A few smaller additions to better address #407:

 - A top level redirect was added to take "docs.bevyengine.org" directly to the `bevy` crate docs.
 - A GitHub Pages CNAME file for supporting a publicly viewable domain instead of `github.io`
 - A robots.txt file is added to disable all search engine crawlers that respect it from crawling it (avoid having conflicting Google search results)
 - A .nojekyl file to speed up deployments since there is no Jekyll templating in the output.

This may require configuration of the  `GITHUB_TOKEN` of the CI to successfully run this.

## Followup

For this to completely resolve #407, a subdomain of https://bevyengine.org/ needs to be set up to point to the CNAME location. This is initially set to "dev-docs.bevyengine.org".

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
This will be very useful. The workflow looks reasonable to me, and it's very clear.

**Comment by Nilirad** [comment]
Great! This will close bevyengine/bevy-website#181.

**Review by mockersf — CHANGES_REQUESTED** [review]
I think it needs to be made very clear in the docs deployed that this would be the doc for Bevy main and not the released version.
Maybe a custom logo with "main" added? Or a sticky header bar that would say "Bevy main - version under development"?

**Comment by Nilirad** [comment]
A message on top could say "Development build docs. Click for the latest release docs", or something like that.

Alternatively, we can have a dropdown menu like docs.rs to choose between development and release docs.

**Comment by james7132** [comment]
Added a sed command that changes the icon out for something more noticeable., and added a header.html shim to force Google not to index, even when linked from other sites. Other search engines should basically stop at the robots.txt.

**Comment by NiklasEi** [comment]
Before more people go looking: the new icon is coming in bevyengine/bevy-website#234

**Comment by james7132** [comment]
With https://github.com/bevyengine/bevy-website/pull/234 merged, @mockersf, any other thing that's needed? It's a little hard to notice right now, but it should be enough to distinguish between it and docs.rs. I could add a sticky banner as suggested earlier, but `--html-before-content` seems to shove it in a row-oriented flexbox due to the navbar. Not familiar enough with HTML/CSS to make it actually stick above everything.

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Pre-docs-build
+        run: |
+          sed -i "s/icon.png/icon-docs-dev.png" src/lib.rs
+          echo "<meta name=\"robots\" content=\"noindex\">" > header.html
+
+      - name: Build docs
+        run: cargo doc --all-features --no-deps -p bevy
+
+      #  This adds the following:
+      #   - A top level redirect to the bevy crate docmentation
```
Comment (mockersf) [inline .github/workflows/docs.yml:38]:
```suggestion
      #   - A top level redirect to the bevy crate documentation
```

**Review by mockersf — APPROVED** [review]

**Comment by mockersf** [comment]
This should be good to go now.

The target domain/url is something that is up to Cart. Maybe it should not live on the website / under a `bevy engine.org` url as we try to keep the website only for the released version? Maybe a subdomain would be enough to differentiate?

I think the url where it would be deployed without custom domain would be https://bevyengine.github.io/bevy, that could help with having website for release, GitHub for in development?

**Comment by cart** [comment]
Yeah I'm ready to move forward on this. I do think that `docs.bevyengine.org` is the wrong name for "in development / unstable" docs. I do think a github url is a reasonable way to correlate the context to "in development", but its worth calling out that github also hosts our "stable releases and branches". We might be able to make it even more clear with a custom url. `unstable-docs.bevyengine.org`, `dev-docs.bevyengine.org`, `dev-api.bevyengine.org` etc.

I think I like the `dev-docs` prefix, for symmetry with docs.rs. I am interested to hear what everyone else thinks.

**Comment by james7132** [comment]
I agree with the `dev-docs` subdomain. It seems recognizable at first glance, shorter than `unstable-docs`, and doesn't look like a REST API subdomain, unlike `dev-api`.

**Comment by cart** [comment]
Cool lets roll with that. I'll get the domain set up.

**Comment by cart** [comment]
We might actually need to merge this pr first. Can you update the domain?

**Pull request closed without merging**